consensus/src/pipeline/pipeline_builder.rs

```diff
@@ -548,15 +566,100 @@ impl PipelineBuilder {
         Ok((Arc::new(sig_verified_txns), block_gas_limit))
     }
 
+    /// Precondition: 1. prepare finishes, 2. parent block's execution phase finishes
+    /// What it does: decides if the block requires a randomness seed and return the value
+    async fn rand_check(
+        prepare_fut: TaskFuture<PrepareResult>,
+        parent_block_execute_fut: TaskFuture<ExecuteResult>,
+        rand_rx: oneshot::Receiver<Option<Randomness>>,
+        executor: Arc<dyn BlockExecutorTrait>,
+        block: Arc<Block>,
+        is_randomness_enabled: bool,
+        rand_check_enabled: bool,
+        module_cache: Arc<Mutex<Option<ValidationState<CachedStateView>>>>,
+    ) -> TaskResult<Option<Randomness>> {
+        let mut tracker = Tracker::start_waiting("rand_check", &block);
+        parent_block_execute_fut.await?;
+        let (user_txns, _) = prepare_fut.await?;
+
+        tracker.start_working();
+        if !is_randomness_enabled {
+            return Ok(None);
+        }
+        let grand_parent_id = block.quorum_cert().parent_block().id();
+        let parent_state_view = executor
+            .state_view(block.parent_id())
+            .map_err(anyhow::Error::from)?;
+
+        let mut has_randomness = false;
+        if rand_check_enabled {
+            let mut cache_guard = module_cache.lock();
+            if let Some(cache_mut) = cache_guard.as_mut() {
+                // flush the cache if the execution state view is not linear
+                // in case of speculative executing a forked block
+                let previous_state_view = cache_mut.state_view_id();
+                let expected_state_view = StateViewId::BlockExecution {
+                    block_id: grand_parent_id,
+                };
+                if previous_state_view == expected_state_view {
+                    cache_mut.reset_state_view(parent_state_view);
+                } else {
+                    cache_mut.reset_all(parent_state_view);
+                }
+            } else {
+                *cache_guard = Some(ValidationState::new(parent_state_view));
+            }
+            let cache_ref = cache_guard.as_mut().expect("just set");
+
+            for txn in user_txns.iter() {
```

Comment:
Should we parallelize this check?

Reply:
this is pretty fast if the data is cached, I have the counter available so if it's too expensive we can parallelize it